src/server/handlers/consents/{ID}/generateChallenge.ts

```diff
@@ -0,0 +1,130 @@
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Abhimanyu Kapur <abhi.kapur09@gmail.com>
+
+ --------------
+ ******/
+
+/* istanbul ignore file */
+
+/*
+ * This flag is to ignore BDD testing
+ * which will be addressed in the future in
+ * ticket #354
+ */
+
+import {
+  updateConsentCredential,
+  isConsentRequestInitiatedByValidSource,
+  putConsentId, ConsentCredential
+} from '../../../../domain/consents/{ID}/generateChallenge'
+import * as challenge from '../../../../lib/challenge'
+import { Request, ResponseToolkit, ResponseObject } from '@hapi/hapi'
+import { consentDB, scopeDB } from '../../../../lib/db'
+import { Consent } from '../../../../model/consent'
+import Logger from '@mojaloop/central-services-logger'
+import { convertScopesToExternal } from '../../../../lib/scopes'
+import { Scope } from '../../../../model/scope'
+
+// Asynchronously deals with generating challenge, updating consent db
+//  and making outgoing PUT consent/{ID} call
+export async function generateChallengeAndPutConsentId (
+  request: Request,
+  consent: Consent,
+  id: string
+): Promise<void> {
+  try {
+    // If there is no pre-existing challenge for the consent id
+    // Generate one and update database
+    if (!consent.credentialChallenge) {
+      // Challenge generation
+      const challengeValue = await challenge.generate()
+
+      // Updating credentials with generated challenge
+      const credential: ConsentCredential = {
+        credentialChallenge: challengeValue,
+        credentialStatus: 'PENDING',
+        credentialType: 'FIDO'
+      }
+
+      consent = await updateConsentCredential(consent, credential)
+    } else if (consent.credentialStatus === 'ACTIVE') {
+      // TODO: Error handling here - dealt with in #355
+      Logger.error('ACTIVE credential consent has requested challenge')
+      return
+    }
+
+    // Retrieve Scopes
+    const scopesRetrieved: Scope[] = await scopeDB.retrieveAll(id)
+    const scopes = convertScopesToExternal(scopesRetrieved)
+
+    // Outgoing call to PUT consents/{ID}
+    await putConsentId(consent, request, scopes)
+  } catch (error) {
+    Logger.push(error)
+    // eslint-disable-next-line max-len
+    Logger.error(`Error: Outgoing call with challenge credential NOT made to  PUT consent/${id}`)
+    // TODO: Decide on error handling HERE -  dealt with in future ticket #355
+  }
+}
+
+/** The HTTP request `POST /consents/{ID}/generateChallenge` is used to create a
+ * credential for the given Consent object. The `{ID}` in the URI should
+ * contain the `{ID}` that was used in the `POST /consents`.
+ * Called by a `PISP`to request a challenge from the `auth-service`, which
+ * will be returned to the PISP via `PUT /consents/{ID}`
+ */
+export async function post (
+  request: Request, h: ResponseToolkit): Promise<ResponseObject> {
+  const id = request.params.id
+
+  // Fetch consent using ID
```

Comment:
according to the discussion we have had yesterday during a standup with @jgeewax and @ramanmangla let move DB related validation to background function: generateChallengeAndPutConsentId and return 202 asap

Reply:
but we want to return a 400 code if request is bad - so that would need to be done synchronously, correct? @ramanmangla  @eoln  @jgeewax

Reply:
if validation is expensive -> if there is a need to do any SQL based query -> we should return 202 ASAP and do rest in background function. If any validation fails we should report about via PUT/.../error . I got the info from @elnyry-sam-k that we can use 6xxxx error codes for PISP: 6x00 - 6x39 dedicated to the API and then rest to the Schemes -> take a look for existing ones: https://github.com/mojaloop/mojaloop-specification/blob/master/documents/Logical%20Data%20Model.md#46-error-codes

Reply:
makes sense will fix this by end of day

Reply:
doen